### PR TITLE
Enforce minimum spend when emergency buying train from a corporation, enforce EBUY_OTHER_VALUE for cashless corporations

### LINKED ISSUE #968: [18Chesapeake] Emergency Fund Raising Rules Inconsistency
When doing emergency fund raising to buy a train from another corporation, the game seems to allow you to sell more shares than necessary to purchase the train. The rules do not permit a player to sell more shares once they have enough funds to buy the train.

I believe the emergency fund raising would work better if it first asked you which train to buy (and at what price if buying from another corporation) and then allowed you to sell shares to reach the necessary funds.

### LINKED ISSUE #2230: President can contribute $1 for cashless corp to buy train from other corp even if EBUY_OTHER_VALUE is false
1846 example: https://www.18xx.games/game/13894?action=376

B&O had $0, and buys a train from PRR for $1, contributed by the B&O's president. The buy trains view at https://www.18xx.games/game/13894?action=375 allows only $1 to be entered, but the PRR's trains should not be listed at all in that scenario.

This affects all titles with `EBUY_OTHER_VALUE` set to false; for currently playable titles, that is just 1846, 18 Los Angeles, and 1889.

### PR DESCRIPTION
**Pull request by michaeljb**
* if shares have been sold to help buy a train, that cash must be used if buying
  from another corporation
* also fix max spend; if the game allows EMR when cross-buying a train, allow
  president to contribute cash even before shares have been sold
* don't allow cashless corporations to ebuy from others when EBUY_OTHER_VALUE is false
* update `validate_all()` in `validate.rb` to take a list of titles

[Fixes #968]
[Fixes #2230]

examples of behavior that this PR fixes: https://www.18xx.games/game/3063?action=442 and https://www.18xx.games/game/13894?action=376

### CODE REVIEW

**Comment by michaeljb** [comment]
These 39 games cannot be migrated, they must be pinned/deleted:

```
[2040, 2121, 2764, 3005, 3191, 3269, 3353, 4411, 5422, 5580, 6348, 6368, 6666, 6959, 7084, 7218, 7219, 7803, 7898, 7923, 8027, 8086, 8402, 8645, 8761, 9179, 9402, 9539, 9584, 9590, 9770, 9808, 10220, 10743, 11969, 12485, 13894, 14099, 14674]
```

#### 18 Los Angeles
14099

#### 1836Jr30
12485
7898
6368
3191
2764

#### 1846
13894

#### 1846 2p Variant
14674
11969

#### 1882
8761
8402
7219
6666

#### 18AL
9590
9584
7803
7218
6348

#### 18Chesapeake
10220
9808
9770
9539
9402
9179
8645
8086
8027
7923
7084
6959
5580
5422
4411
3353
3269
3005
2121
2040

#### 18TN
10743

There are 34 more games that are broken by the changes on this branch, but they're not raising the `corporation may not spend...` error, so I'm still investigating those.

**Comment by michaeljb** [comment]
> There are 34 more games that are broken by the changes on this branch, but they're not raising the `corporation may not spend...` error, so I'm still investigating those.

79539242 fixes most of those, just 12 left, all 1817....

**Comment by michaeljb** [comment]
> just 12 left, all 1817...

Actually, all 12 of those games have been deleted on the live site, this PR is ready to go :tada:

File: lib/engine/step/buy_train.rb [hunk 1]
```diff
@@ -38,6 +38,18 @@ def pass!
       end
 
       def process_buy_train(action)
+        if action.train.owned_by_corporation?
+          min, max = spend_minmax(action.entity, action.train)
+          unless Range.new(min, max).include?(action.price)
```
Comment (tobymao) [inline lib/engine/step/buy_train.rb:43]: (min..max).include?

**Review by tobymao — APPROVED** [review]

**Comment by tobymao** [comment]
feel free to merge and migrate yourself

**Pull request merged**